services/redis.go

```diff
@@ -18,79 +18,85 @@ package services
+// RedisClient is a local type which embeds the imported redis.Client to include its own functionality
 type RedisClient struct {
-	Client *redisV9.Client
-}
-
-// explicit checks for config params are done because the go-redis package lets us create a client
-// with incorrect params, so errors are only found during subsequent command executions
-func createRedisClient(conf RedisConfiguration) (*redisV9.Client, error) {
-	if conf.RedisEndpoint == "" {
-		err := errors.New("Redis server address must not be empty")
-		log.Error().Err(err)
-		return nil, err
-	}
-
-	if conf.RedisDatabase < 0 || conf.RedisDatabase > 15 {
-		err := errors.New("Redis selected database must be a value in the range 0-15")
-		log.Error().Err(err)
-		return nil, err
-	}
-
-	log.Info().Msgf("creating redis client. endpoint %v, selected DB %d, timeout seconds %d",
-		conf.RedisEndpoint, conf.RedisDatabase, conf.RedisTimeoutSeconds,
-	)
-
-	// DB is configurable in case we want to change data structure
-	c := redisV9.NewClient(&redisV9.Options{
-		Addr:        conf.RedisEndpoint,
-		DB:          conf.RedisDatabase,
-		Password:    conf.RedisPassword,
-		ReadTimeout: time.Duration(conf.RedisTimeoutSeconds) * time.Second,
-	})
-
-	return c, nil
+	redis.Client
 }
 
 // NewRedisClient creates a new Redis client based on configuration and returns RedisInterface
 func NewRedisClient(conf RedisConfiguration) (RedisInterface, error) {
-	client, err := createRedisClient(conf)
+	client, err := redis.CreateRedisClient(
+		conf.RedisEndpoint,
+		conf.RedisDatabase,
+		conf.RedisPassword,
+		conf.RedisTimeoutSeconds,
+	)
 	if err != nil {
 		return nil, err
 	}
 
 	return &RedisClient{
-		Client: client,
+		redis.Client{Connection: client},
 	}, nil
 }
 
-// HealthCheck executes PING command to check for liveness status of Redis server
-func (redis *RedisClient) HealthCheck() (err error) {
+// GetRequestIDsForClusterID retrieves a list of request IDs from Redis.
+// "List" of request IDs is in the form of keys with empty values in the following structure:
+// organization:{org_id}:cluster:{cluster_id}:request:{request_id1}.
+func (redis *RedisClient) GetRequestIDsForClusterID(
+	orgID types.OrgID,
+	clusterID types.ClusterName,
+) (requestIDs []types.RequestID, err error) {
 	ctx := context.Background()
 
-	// .Result() gets value and error of executed command at once
-	res, err := redis.Client.Ping(ctx).Result()
-	if err != nil || res != "PONG" {
-		log.Error().Err(err).Msg("Redis PING command failed")
-		return errors.New(redisExecutionFailedMsg)
+	scanKey := fmt.Sprintf(RequestIDsScanPattern, orgID, clusterID)
+
+	var cursor uint64
+	for {
+		var keys []string
+		var err error
+		keys, cursor, err = redis.Client.Connection.Scan(ctx, cursor, scanKey, 0).Result()
+		if err != nil {
+			log.Error().Err(err).Msgf("failed to execute SCAN command for key '%v' and cursor '%d'", scanKey, cursor)
+			return nil, err
+		}
+
+		// get last part of key == request_id
+		for _, key := range keys {
```

Comment:
probably more efficient to initialise the slice and insert with indexing.

                requestIDs = make(..., len(keys), len(keys))
		for idx, key := range keys {
			keySliced := strings.Split(key, ":")
			requestID := keySliced[len(keySliced)-1]
			requestIDs[idx] = types.RequestID(requestID)
		}

Reply:
@epapbak we don't know the total length though, SCAN doesn't return the whole dataset, it returns a "page" with a default length of 10 and it returns a cursor which needs to be used in subsequent SCAN executions to get the whole dataset, hence the while loop.

Reply:
+ it does not run in transaction, so it is not possible to tell the exact number of records (AFAIK - dunno if/how Redis will work)